pyadjoint/optimization/tao_solver.py

```diff
@@ -0,0 +1,361 @@
+            raise TypeError("MinimizationProblem required")
+        if problem.constraints is not None:
+            raise NotImplementedError("Constraints not implemented")
+
+        if comm is None:
+            comm = PETSc.COMM_WORLD
+        if hasattr(comm, "tompi4py"):
+            comm = comm.tompi4py()
+
+        taoobjective = TAOObjective(problem.reduced_functional)
+
+        vec_interface = PETScVecInterface(
+            tuple(m.control for m in taoobjective.reduced_functional.controls),
+            comm=comm)
+        n, N = vec_interface.n, vec_interface.N
+        to_petsc, from_petsc = vec_interface.to_petsc, vec_interface.from_petsc
+
+        tao = PETSc.TAO().create(comm=comm)
+
+        def objective_gradient(tao, x, g):
+            M = taoobjective.new_M()
+            from_petsc(x, M)
+            J_val, dJ = taoobjective.objective_gradient(M)
+            to_petsc(g, dJ)
+            return J_val
+
+        tao.setObjectiveGradient(objective_gradient, None)
+
+        def hessian(tao, x, H, P):
+            H.getPythonContext().set_M(x)
+
+        class Hessian:
+            def __init__(self):
+                self._shift = 0.0
+
+            @cached_property
+            def _M(self):
+                return taoobjective.new_M()
+
+            def set_M(self, x):
+                from_petsc(x, self._M)
+                self._shift = 0.0
+
+            def shift(self, A, alpha):
+                self._shift += alpha
+
+            def mult(self, A, x, y):
+                M_dot = taoobjective.new_M()
+                from_petsc(x, M_dot)
+                ddJ = taoobjective.hessian(self._M, M_dot)
+                to_petsc(y, ddJ)
+                if self._shift != 0.0:
+                    y.axpy(self._shift, x)
+
+        H_matrix = PETSc.Mat().createPython(((n, N), (n, N)),
+                                            Hessian(), comm=comm)
+        H_matrix.setOption(PETSc.Mat.Option.SYMMETRIC, True)
+        H_matrix.setUp()
+        tao.setHessian(hessian, H_matrix)
+
+        class GradientNorm:
+            def mult(self, A, x, y):
+                X = taoobjective.new_M_dual()
+                from_petsc(x, X)
+                assert len(taoobjective.reduced_functional.controls) == len(X)
+                X = tuple(m._ad_convert_type(x, {"riesz_representation": inner_product})
+                          for m, x in zip(taoobjective.reduced_functional.controls, X))
+                to_petsc(y, X)
+
+        M_inv_matrix = PETSc.Mat().createPython(((n, N), (n, N)),
+                                                GradientNorm(), comm=comm)
+        M_inv_matrix.setOption(PETSc.Mat.Option.SYMMETRIC, True)
+        M_inv_matrix.setUp()
+        tao.setGradientNorm(M_inv_matrix)
+
+        if problem.bounds is not None:
+            def convert_bound(m, b, default):
+                if b is None:
+                    b = default
+                return m._ad_convert_type(b)
```

Comment:
When using the method `m._ad_convert_type(b)`, do you expect to be returned a `Function` to have its data array equal to `b`?

Reply:
I believe here the bound value `b` should either have the same type as `m`, or be a scalar, yes. What is needed is "`m` dofs = `b`", but I don't see an easy way to do that via `OverloadedType` interfaces.

Reply:
I think `_ad_convert_type` should have more options besides just `riesz_representation`. Perhaps something like `m._ad_convert_type(b, options={"convert_to_function": True})` could be useful. I will think more about this and discuss it with David. I will keep you updated.

Reply:
I've removed the use of `_ad_convert_type` here. If there is to be a check on bounds type then I think that should be in `OptimizationProblem.__check_arguments` -- at the moment this allows `int` or `float` bounds independent of the control type.